### ISSUE TITLE
[CartesianChart] SelectionBehavior.selectDataPoints not working always

### ISSUE THREAD

**Issue by andrea689** [syncfusion/flutter-widgets#86]
Hi, I want select a point when user tap on label, so I use:

```dart
                onDataLabelTapped: (onTapArgs) {
                  try {
                    selectionBehavior.selectDataPoints(onTapArgs.overallPointIndex);
                  } catch (e) {
                    selectionBehavior.selectDataPoints(onTapArgs.pointIndex);
                  }
                },
```

I use try/catch because sometime `selectDataPoints` returns `RangeException`, I think that it use only visible points in fact `pointIndex` works correctly.

**Comment by SriramKiranSenthilkumar** [comment]
Hi @andrea689,

Greetings from Syncfusion. We have analyzed your scenario at our end and we would like to share some information regarding the **overallPointIndex** argument of the onDataLabelTapped event available in the chart widget. The **overallPointIndex** argument available in the data label tapped callback event can be used as an information to know about the overall point index of a data point when its data label is tapped. Also to mention that the **selectDataPoints()** public method of the **SelectionBehavior** can only be used to select the data points that are in the visible range and so we kindly request you to use the **pointIndex** argument to select the data points in the chart when tapping on its data label in the onDataLabelTapped event. 

Also, regarding this query " sometime selectDataPoints returns RangeException", we tried to replicate the reported issue at our end regarding the **selectDataPoints** method returned range exception in the chart. Unfortunately, we are not able to replicate the issue so we kindly request you to revert us with a simple replication sample to replicate the reported exception at our end so that it will be helpful in further analysis in order to provide the solutions sooner.

Regards,
Sriram Kiran.

**Comment by andrea689** [comment]
<details>
  <summary>This is a code example</summary>

```dart
import 'package:flutter/material.dart';
import 'package:syncfusion_flutter_charts/charts.dart';

void main() {
  return runApp(ChartApp());
}

class ChartApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Chart Demo',
      theme: ThemeData(primarySwatch: Colors.blue, brightness: Brightness.dark),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  // ignore: prefer_const_constructors_in_immutables
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    final selectionBehavior = SelectionBehavior(
      enable: true,
    );

    return Scaffold(
      appBar: AppBar(
        title: const Text('Syncfusion Flutter chart'),
      ),
      body: Container(
        height: 300,
        child: SfCartesianChart(
          primaryXAxis: CategoryAxis(
            visibleMinimum: 17,
            visibleMaximum: 23,
          ),
          primaryYAxis: NumericAxis(),
          zoomPanBehavior: ZoomPanBehavior(
            enablePinching: true,
            enablePanning: true,
            zoomMode: ZoomMode.x,
          ),
          series: <ChartSeries<SalesData, DateTime>>[
            ColumnSeries<SalesData, DateTime>(
              selectionBehavior: selectionBehavior,
              dataSource: chartData,
              xValueMapper: (SalesData sales, _) => sales.year,
              yValueMapper: (SalesData sales, _) => sales.sales,
              dataLabelSettings: DataLabelSettings(
                isVisible: true,
                builder: (data, point, series, pointIndex, seriesIndex) {
                  final datum = data as SalesData;
                  return Container(
                    height: 36,
                    child: Column(
                      children: [
                        Icon(Icons.note, size: 18),
                        Text(datum.sales.toString()),
                      ],
                    ),
                  );
                },
              ),
            )
          ],
          onSelectionChanged: (selectionArgs) {
            print(chartData[selectionArgs.overallDataPointIndex].sales);
          },
          onDataLabelTapped: (onTapArgs) {
            /// [selectDataPoints] calls [onSelectionChanged]
            print('ASD ${onTapArgs.overallPointIndex} ${onTapArgs.pointIndex}');
            try {
              selectionBehavior.selectDataPoints(onTapArgs.overallPointIndex);
            } catch (e) {
              selectionBehavior.selectDataPoints(onTapArgs.pointIndex);
            }
          },
        ),
      ),
    );
  }
}

class SalesData {
  SalesData(this.year, this.sales);

  final DateTime year;
  final double sales;
}

final chartData = [
  SalesData(DateTime(2019, 6, 10), 3000),
  SalesData(DateTime(2019, 6, 11), 800),
  SalesData(DateTime(2019, 6, 12), 300),
  SalesData(DateTime(2019, 6, 13), 500),
  SalesData(DateTime(2019, 6, 14), 500),
  SalesData(DateTime(2019, 6, 15), 400),
  SalesData(DateTime(2019, 6, 16), 800),
  SalesData(DateTime(2019, 6, 17), 200),
  SalesData(DateTime(2019, 6, 18), 800),
  SalesData(DateTime(2019, 6, 19), 500),
  SalesData(DateTime(2019, 6, 20), 300),
  SalesData(DateTime(2019, 6, 21), 1200),
  SalesData(DateTime(2019, 6, 22), 200),
  SalesData(DateTime(2019, 6, 23), 500),
  SalesData(DateTime(2019, 6, 24), 800),
  SalesData(DateTime(2019, 6, 25), 700),
  SalesData(DateTime(2019, 6, 26), 0),
  SalesData(DateTime(2019, 6, 27), 100),
  SalesData(DateTime(2019, 6, 28), 600),
  SalesData(DateTime(2019, 6, 29), 3500),
  SalesData(DateTime(2019, 6, 30), 700),
  SalesData(DateTime(2019, 7, 1), 500),
  SalesData(DateTime(2019, 7, 2), 300),
  SalesData(DateTime(2019, 7, 3), 800),
  SalesData(DateTime(2019, 7, 4), 300),
  SalesData(DateTime(2019, 7, 5), 200),
  SalesData(DateTime(2019, 7, 6), 500),
  SalesData(DateTime(2019, 7, 7), 900),
  SalesData(DateTime(2019, 7, 8), 200),
  SalesData(DateTime(2019, 7, 9), 400),
  SalesData(DateTime(2019, 7, 10), 100),
  SalesData(DateTime(2019, 7, 11), 800),
  SalesData(DateTime(2019, 7, 12), 600),
  SalesData(DateTime(2019, 7, 13), 3200),
  SalesData(DateTime(2019, 7, 14), 1000),
  SalesData(DateTime(2019, 7, 15), 800),
];

```

</details>

if I don't use `onSelectionChanged` the `RangeException` not thrown, but the selection not working first time, but after move the chart it working.

So, if I don't pan the chart, `overallPointIndex` working.. After pan the chart, `pointIndex` is working..

**Comment by SriramKiranSenthilkumar** [comment]
Hi @andrea689,

Thanks for the revert. We have analyzed your scenario with the provided code snippet and we are able to replicate the reported issue at our end regarding the selection with the public method using the overallPointIndex is not working when the chart is set in the visible range. We will consider this issue as a bug and will include the fix for the reported issue in our upcoming weekly patch release which is expected to be rolled out by 5th January 2020. We will update you once the release is rolled out and we appreciate your patience until then.

Also, we would like to let you know that, in our latest chart widget version([18.4.31](https://pub.dev/packages/syncfusion_flutter_charts/versions/18.4.31)), the **pointIndex** argument which indicates the data point index according to the chart's visible range/viewport is renamed as **viewportPointIndex** and the **overallPointIndex** argument which indicates the data point index according to the whole chart data source is renamed as **pointIndex**.

Regards,
Sriram Kiran

**Comment by Dharanipalani** [comment]
Hi @andrea689,

We are glad to inform you that the reported issue has been fixed in our latest version v18.4.33. Now, we made the changes that all types of selection will work with pointIndex property, so hereafter use the pointIndex for selection in onSelectionChanged event. Kindly upgrade to our latest version to resolve the reported scenario and it can get from below,

Latest version: [https://pub.dev/packages/syncfusion_flutter_charts/versions/18.4.33]

Thanks,
Dharanitharan. P
